### DISCUSSION TITLE
More relationship questions

### DISCUSSION THREAD

**Discussion by the1gofer** [sqlalchemy/sqlalchemy#6721] [Usage Questions] [answered]
I'm sure I'm missing something simple, and hope you get point me in the right direction. 

Error message:
`One or more mappers failed to initialize - can't proceed with initialization of other mappers. Triggering mapper: 'mapped class Source->sources'. Original exception was: Could not determine join condition between parent/child tables on relationship Source.similar_sources - there are multiple foreign key paths linking the tables.  Specify the 'foreign_keys' argument, providing a list of those columns which should be counted as containing a foreign key reference to the parent table.`

Source:
```
class Source(Base):
    """table to store sources (obituaries, headstones, etc)"""

    __tablename__ = 'sources'

    id = Column(Integer, primary_key=True)
    title = Column(String)
    text = Column(Text)
    accessed_date_id = Column(Integer, ForeignKey('dates.id'))
    publication_date_id = Column(Integer, ForeignKey('dates.id'))
    type = Column(String, nullable=False)
    wiki_page = Column(Text)
    wikified = Column(Boolean, default=False)
    url_id = Column(Integer, ForeignKey('urls.id'), nullable=False)
    primary_person_id = Column(Integer, ForeignKey('people.id'))

    # relationships
    primary_person = relationship('Person')
    files = relationship('File', back_populates="source")
    accessed_date = relationship('Date', foreign_keys=[accessed_date_id])
    publication_date = relationship('Date', foreign_keys=[publication_date_id])
    url = relationship('URL', foreign_keys=[url_id], back_populates="source", uselist=False)
    people = relationship('Person', secondary=people_source_table, back_populates="sources")
    locations = relationship('Location', secondary=location_source_table, back_populates="sources")
    dates = relationship('Date', secondary=date_source_table, back_populates="sources")
    similar_sources = relationship("SimilarSources", back_populates="Source")
```

SimilarSources:
```
class SimilarSources(Base):
    """table to store obit similarity so we don't have to spend time running the slow script every time"""

    __tablename__ = "similar_sources"
    id = Column(Integer, primary_key=True)
    source_one_id = Column(Integer, ForeignKey('sources.id'), nullable=False)
    source_two_id = Column(Integer, ForeignKey('sources.id'), nullable=False)
    similarity = Column(Float)

    source_one = relationship('Source', back_populates="similar_sources", foreign_keys=[source_one_id])
    source_two = relationship('Source', back_populates="similar_sources", foreign_keys=[source_two_id])
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

I think the error message has all the information you need for this. Let's break it down:
> One or more mappers failed to initialize - can't proceed with initialization of other mappers. Triggering mapper: 'mapped class Source->sources'. Original exception was: Could not determine join condition between parent/child tables on relationship Source.similar_sources 

The issue is in `similar_sources = relationship("SimilarSources", back_populates="Source")`

> there are multiple foreign key paths linking the tables. 

There are indeed multiple fk from `sources` to `similar_sources`:
```
source_one_id = Column(Integer, ForeignKey('sources.id'), nullable=False)
source_two_id = Column(Integer, ForeignKey('sources.id'), nullable=False)
```

> Specify the 'foreign_keys' argument, providing a list of those columns which should be counted as containing a foreign key reference to the parent table.

The suggested resolution is to specify which one of the two fk you want to track with the relationship using the `foreign_keys` parameter in the relationship.

**Reply by the1gofer** [reply]
Thanks.  Isn't that what i've done here?

source_one = relationship('Source', back_populates="similar_sources", foreign_keys=[source_one_id])
source_two = relationship('Source', back_populates="similar_sources", foreign_keys=[source_two_id])

**Reply by zzzeek** [reply]
> Thanks. Isn't that what i've done here?
> 
> source_one = relationship('Source', back_populates="similar_sources", foreign_keys=[source_one_id])
> source_two = relationship('Source', back_populates="similar_sources", foreign_keys=[source_two_id])

yes but the error message refers to the "similar_sources" relationship, you would need foreign_keys in those relationships() as well

**Reply by the1gofer** [reply]
Ok, I think I have multiple issues going on. 

I decided it doessn't need to be bidirectional.  so I've got 
`similar_sources_id = Column(Integer, ForeignKey('similar_sources.id')) `
and 
```
__tablename__ = "similar_sources"
    id = Column(Integer, primary_key=True)
    source_one_id = Column(Integer, ForeignKey('sources.id'), nullable=False)
    source_two_id = Column(Integer, ForeignKey('sources.id'), nullable=False)
    similarity = Column(Float)

    source_one = relationship('Source',foreign_keys=[source_one_id])
    source_two = relationship('Source',foreign_keys=[source_two_id])
```

I can't get it to create the similar_sources_id on the source table.  Normally, I just use  Base.metadata.drop_all(bind=engine) and Base.metadata.create_all(bind=engine) to get a fresh restart.  However, it gives me:`sqlalchemy.exc.CircularDependencyError: Can't sort tables for DROP; an unresolvable foreign key dependency exists between tables: similar_sources, sources.  Please ensure that the ForeignKey and ForeignKeyConstraint objects involved in the cycle have names so that they can be dropped using DROP CONSTRAINT.`

**Reply by zzzeek** [reply]
you have a cycle between the foreign key constraints on both tables.    Give the newly added FK constraint a name ( `ForeignKey("foo.id", name="my_fkey")` ) so that SQLAlchemy can add this constraint after the two tables are created.   for the drop, you probably should drop manually from your DB just the once as this model doesn't match what you have in the DB.

**Reply by the1gofer** [reply]
would you use the same name for every item that uses foo.id, or make a different name for each?

**Reply by CaselIT** [reply]
I don't think the db support duplicate names. IIRC postgresql for example will error if two objects have the same name.
